cpp/src/Glacier2/SessionRouterI.cpp

```diff
@@ -320,136 +220,108 @@ class SSLCreateSession : public CreateSession
     {
         if(ok)
         {
-            authorized(_sessionRouter->_sslSessionManager);
+            authorized(_sessionRouter->_sslSessionManager != nullptr);
         }
         else
         {
-            exception(PermissionDeniedException(reason.empty() ? string("permission denied") : reason));
+            exception(make_exception_ptr(PermissionDeniedException(reason.empty() ? string("permission denied") : reason)));
         }
     }
 
     void
-    authorizeException(const Ice::Exception& ex)
+    authorizeException(exception_ptr ex)
     {
-        if(dynamic_cast<const PermissionDeniedException*>(&ex))
+        try
+        {
+            rethrow_exception(ex);
+        }
+        catch(const PermissionDeniedException&)
         {
             exception(ex);
         }
-        else
+        catch(...)
         {
             unexpectedAuthorizeException(ex);
         }
     }
 
-    virtual void
-    authorize()
+    void
+    authorize() override
     {
         assert(_sessionRouter->_sslVerifier);
 
-        Ice::Context ctx = _current.ctx;
+        auto ctx = _current.ctx;
         ctx.insert(_context.begin(), _context.end());
-        _sessionRouter->_sslVerifier->begin_authorize(_sslInfo, ctx,
-                                                      newCallback_SSLPermissionsVerifier_authorize(this,
-                                                                             &SSLCreateSession::authorizeResponse,
-                                                                             &SSLCreateSession::authorizeException));
+
+        auto self = static_pointer_cast<SSLCreateSession>(shared_from_this());
+        _sessionRouter->_sslVerifier->authorizeAsync(_sslInfo,
+                                                     [self](bool ok, const string& reason)
+                                                     {
+                                                         self->authorizeResponse(ok, reason);
```

Comment:
I would get rid of the helper functions called only once from AMI callbacks:
authorizeResponse|Exception,  checkPermissionsResponse|Exception